homeassistant/components/growatt_server/sensor.py

```diff
@@ -0,0 +1,123 @@
+from homeassistant.const import CONF_NAME
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_USERNAME = "username"
+CONF_PASSWORD = "password"
+CONF_PLANT_ID = "plant_id"
+DEFAULT_UNIT = "kW"
+DEFAULT_NAME = "Growatt"
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PLANT_ID, default="0"): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Growatt sensor."""
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    plant_id = config.get(CONF_PLANT_ID)
+    name = config.get(CONF_NAME)
+
+    api = growattServer.GrowattApi()
+
+    # Log in to api and fetch first plant if no plant id is defined.
+    user_id = api.login(username, password)["userId"]
+    if plant_id == "0":
+        plant_info = api.plant_list(user_id)
+        plant_id = plant_info["data"][0]["plantId"]
+        pass
+
+    # Get a list of inverters for specified plant to add sensors for.
+    inverters = api.inverter_list(plant_id)
+
+    # Add a sensor for the total of the whole plant.
+    add_devices(
+        [GrowattInverter(api, name + "_Total", plant_id, username, password)], True
+    )
+
+    # Add sensors for each inverter in the specified plant.
+    for inverter in inverters:
+        add_devices(
+            [
+                GrowattInverter(
+                    api,
+                    name + "_" + inverter["deviceAilas"],
+                    inverter["deviceSn"],
+                    username,
+                    password,
+                )
+            ],
+            True,
+        )
+
+
+class GrowattInverter(Entity):
+    """Representation of a Growattt Sensor."""
+
+    def __init__(self, api, name, inverter_id, u, p):
+        """Initialize a PVOutput sensor."""
+        self.api = api
+        self._name = name
+        self.inverter_id = inverter_id
+        self.username = u
+        self.password = p
+        self.is_total = "Total" in name and inverter_id.isdigit()
+        self._state = None
+        self.attributes = {}
+        self._unit_of_measurement = "W"
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return self._unit_of_measurement
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the monitored installation."""
+        attributes = self.attributes
```

Comment:
Please set explicitly the items in the attributes. We want to see what they are and check formatting.

Reply:
It is quite a list and depending on the person, different pieces of it may be used.
For instance i use pvv1 and pvv2 to read the wattage of my two strands connected to it.
While another person used `plantMoneyText` from the total sensor to see how much money he saved by using them this month.
![image](https://user-images.githubusercontent.com/15870933/63641261-a5825480-c6ab-11e9-92aa-55f78964cbec.png)
Do you want me to filter it so you know whats in there or to translate what it returns to a more readable format?

Reply:
I'd probably translate the keys to more readable format if there's no strong argument for keeping the names.

We want to all of them explicitly det, so we know what kind of values are included. We have some special rules eg regarding formatting, lowercase snakecase, and time values, only absolute utc, that we want to make sure are followed.

Reply:
I have made them more readable and left the ones i don't use and don't have a value out since i don't know what they are.